src/adapter/src/catalog/open.rs

```diff
@@ -852,51 +805,60 @@ impl Catalog {
 
-        let last_seen_version = catalog
-            .storage()
-            .await
+        let last_seen_version = txn
             .get_catalog_content_version()
-            .await?
             .unwrap_or_else(|| "new".to_string());
 
         if !config.skip_migrations {
-            migrate::migrate(&catalog, config.connection_context)
+            migrate::migrate(&mut state, &mut txn, config.now, config.connection_context)
                 .await
                 .map_err(|e| {
                     Error::new(ErrorKind::FailedMigration {
                         last_seen_version: last_seen_version.clone(),
-                        this_version: catalog.config().build_info.version,
+                        this_version: config.build_info.version,
                         cause: e.to_string(),
                     })
                 })?;
-            catalog
-                .storage()
-                .await
-                .set_catalog_content_version(catalog.config().build_info.version)
-                .await?;
+            txn.set_catalog_content_version(config.build_info.version.to_string())?;
         }
 
-        let mut catalog = {
-            let mut storage = catalog.storage().await;
-            let mut tx = storage.transaction().await?;
-            let catalog = Self::load_catalog_items(&mut tx, &catalog)?;
-            tx.commit().await?;
-            catalog
+        let mut state = Catalog::load_catalog_items(&mut txn, &state)?;
+
+        let mut builtin_migration_metadata = Catalog::generate_builtin_migration_metadata(
+            &state,
+            &mut txn,
+            migrated_builtins,
+            id_fingerprint_map,
+        )?;
+        Catalog::apply_in_memory_builtin_migration(&mut state, &mut builtin_migration_metadata)?;
+        Catalog::apply_persisted_builtin_migration(
+            &state,
+            &mut txn,
+            &mut builtin_migration_metadata,
+        )?;
+
+        txn.commit().await?;
+
+        let mut catalog = Catalog {
+            state,
+            plans: CatalogPlans {
+                optimized_plan_by_id: Default::default(),
+                physical_plan_by_id: Default::default(),
+                dataflow_metainfos: BTreeMap::new(),
+            },
+            transient_revision: 1,
+            storage: Arc::new(tokio::sync::Mutex::new(storage)),
         };
 
-        let mut builtin_migration_metadata = catalog
-            .generate_builtin_migration_metadata(migrated_builtins, id_fingerprint_map)
-            .await?;
-        catalog.apply_in_memory_builtin_migration(&mut builtin_migration_metadata)?;
+        // After LD is loaded, set the intended stash timeout.
+        // TODO: Move this into the stash constructor.
         catalog
-            .apply_persisted_builtin_migration(&mut builtin_migration_metadata)
-            .await?;
+            .storage()
+            .await
+            .set_connect_timeout(catalog.system_config().crdb_connect_timeout())
```

Comment:
Do you think there is any harm in setting this after such a large commit? AFAICT it's tricky to set it before the commit because we don't have LD yet, so I don't think we can necessarily move it, but just wanted to call it out in case we want to hard code a higher timeout for this initial commit

Reply:
CC @mjibson, do you remember what the motivation for the CRDB timeout was? 

If the issue was that the timeout was too low then we might be in trouble because we may find ourselves in a situation where open is never able to complete in time. On the other hand if the timeout was too high, then I think that's fine. If open would have timed out then we couldn't have started anyway. So I think it's OK to wait a little bit longer for the initial commit transaction.

Also this value was always set after the transaction where we initialize the catalog for new environments, so that's more evidence that this is probably OK.

Reply:
Another point is that the general case is there are no modifications during `Catalog::open`. Only if there's migrations will there be some modifications to make. 

I'm going to go ahead and merge this since it's blocking some other things, and if the timeout is a problem I can revisit/revert.

Reply:
https://github.com/MaterializeInc/materialize/pull/17852 was the motivation for the timeout, which is needed during cockroach rolling restarts.

Reply:
I think setting this after doing a bunch of other stuff means that in the worse case, envd boot might take a while longer (~2 minutes longer waiting for the default timeout) if we are booting during a cockroach restart. This isn't great but also not terrible. We really really want the steady state operation to be fast, which this still achieves. I think it'd be nice to have this set before doing any operation on cockroach, but we have some bootstrap ordering problems there with LD and persisted values iirc.

We could have a hard coded initial timeout value that would protect against this problem during boot, then this gets applied from LD later.

Reply:
Ok, got it. I think the initial timeout is a good idea. I'll whip something up by EOD.

Reply:
This should restore it to it's old behavior: https://github.com/MaterializeInc/materialize/pull/22802